tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +86,281 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
+                'HTML' => '<p class="foo">hello</p><p class="foo">world</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '1 attribute, 2 classes, both to be kept' => [
+                'HTML' => '<p class="foo bar">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+        ];
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function provideHtmlWithExtraWhitespaceAndAllMatchedClasses()
+    {
+        return [
+            '1 attribute, 1 class with extra whitespace, that class to be kept' => [
+                'HTML' => '<p class=" foo ">hello</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '1 attribute, 2 classes with extra whitespace, both to be kept' => [
+                'HTML' => '<p class=" foo  bar ">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, both to be kept' => [
+                'HTML' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo', 'bar'],
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string[] $classesToKeep
+     *
+     * @dataProvider provideHtmlAndSomeMatchedClasses
+     * @dataProvider provideHtmlAndAllMatchedClasses
+     * @dataProvider provideHtmlWithExtraWhitespaceAndSomeMatchedClasses
+     * @dataProvider provideHtmlWithExtraWhitespaceAndAllMatchedClasses
+     */
+    public function removeRedundantClassesNotRemovesClassesToKeep($html, array $classesToKeep)
+    {
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        $result = $subject->render();
+        foreach ($classesToKeep as $class) {
+            $expectedInstanceCount = \substr_count($html, $class);
+            self::assertSame(
+                $expectedInstanceCount,
+                \substr_count($result, $class),
+                'asserting \'' . $result . '\' contains ' . $expectedInstanceCount . ' instance(s) of "' . $class
+                . '"'
+            );
+        }
+    }
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string[] $classesToKeep
+     *
+     * @dataProvider provideHtmlWithExtraWhitespaceAndSomeMatchedClasses
+     * @dataProvider provideHtmlWithExtraWhitespaceAndAllMatchedClasses
+     */
+    public function removeRedundantClassesMinifiesClassAttributes($html, array $classesToKeep)
+    {
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        \preg_match_all('/class="([^"]*+)"/', $subject->render(), $classAttributeMatches);
+        foreach ($classAttributeMatches[1] as $classAttributeValue) {
+            self::assertNotRegExp('/^\\s|\\s{2}|\\s$/', $classAttributeValue);
```

Comment:
Can we move this into a static `assert*` method with a descriptive name?

Reply:
Done.